### DISCUSSION TITLE
CosineSimilarity code Explain

### DISCUSSION THREAD

**Discussion by mtsol** [NVIDIA/spark-rapids-examples#281] [General]
Hello team, I need to know the function call at line 34 in https://github.com/NVIDIA/spark-rapids-examples/blob/branch-22.12/examples/UDF-Examples/RAPIDS-accelerated-UDFs/src/main/java/com/nvidia/spark/rapids/udf/java/CosineSimilarity.java executes on GPU or CPU? And if it executes on CPU then where is the GPU code?

**Comment by nvliyuan** [comment]
Hi @mtsol , this executes on GPU, please refer this [doc](https://nvidia.github.io/spark-rapids/docs/additional-functionality/rapids-udfs.html#adding-gpu-implementations-to-udfs)

**Comment by mtsol** [comment]
Do you have any such example in scala?

**Comment by nvliyuan** [comment]
please find this url [encode&decode](https://github.com/NVIDIA/spark-rapids-examples/tree/main/examples/UDF-Examples/RAPIDS-accelerated-UDFs/src/main/scala/com/nvidia/spark/rapids/udf/scala) example
